src/wasm/wasm-type.cpp

```diff
@@ -384,6 +475,14 @@ HeapTypeInfo::~HeapTypeInfo() {
   WASM_UNREACHABLE("unexpected kind");
 }
 
+bool HeapTypeInfo::getCanonical(HeapType& out) const {
+  if (isFinalized && kind == BasicKind) {
```

Comment:
why does it matter if it's finalized or not?

Reply:
Imagine I have a temporary heap type backed by a HeapTypeInfo in a TypeBuilder. It may currently be set to a BasicHeapType, but in the future I may change it to be some other complex HeapType instead. It would be bad if temporary Types that refer to that TypeBuilder slot were prematurely canonicalized on creation to refer directly to a BasicHeapType instead because they would not observe the later changes to the temporary (mutable) HeapType.

Reply:
I may have forgotten something. What is the use case for having a type in a TypeBuilder change? I thought the usage is to define the N types, each once, and then the builder handles canonicalization etc., but none of that could replace a basic heap type with a complex one?

Reply:
There is no concrete use case specifically for replacing a basic heap type with a complex one, but we do depend more generally on the type being able to be set more than once because the TypeBuilder entries are initialized to arbitrary types and then set again by the user later. So far we have not enforced any additional constraints on how it can be mutated. IMO, the simplest mental model for the TypeBuilder is that it is entirely mutable with no constraints.

Reply:
I see, that makes sense. It does seem hard to think of an API that would not allow user changes.